nix/wire-server.nix

```diff
@@ -0,0 +1,271 @@
+      galley-integration= [pkgs.mls-test-cli];
+    };
+
+    images = localMods@{enableOptimization, enableDocs, enableTests}:
+      attrsets.mapAttrs (execName: drv:
+        pkgs.dockerTools.buildLayeredImage {
+          name = "quay.io/wire/${execName}";
+          maxLayers = 10;
+          contents = [
+            pkgs.cacert
+            pkgs.iana-etc
+            pkgs.coreutils
+            pkgs.bashInteractive
+            pkgs.dumb-init
+            drv
+            tmpDir
+          ] ++ pkgs.lib.optionals (builtins.hasAttr execName extraContents) (builtins.getAttr execName extraContents);
+          fakeRootCommands = ''
+                           chmod 1777 tmp
+                           chmod 1777 var/tmp
+                           '';
+          config = {
+            Entrypoint = ["${pkgs.dumb-init}/bin/dumb-init" "--" "${drv}/bin/${execName}"];
+          };
+        }
+      ) (staticExecs localMods);
+
+    localModsEnableAll = {
+      enableOptimization = true;
+      enableDocs = true;
+      enableTests = true;
+    };
+    localModsOnlyTests = {
+      enableOptimization = false;
+      enableDocs = false;
+      enableTests = true;
+    };
+    localModsOnlyDocs = {
+      enableOptimization = false;
+      enableDocs = true;
+      enableTests = false;
+    };
+    imagesList = pkgs.writeTextFile {
+      name = "imagesList";
+      text = "${lib.concatStringsSep "\n" (builtins.attrNames (images localModsEnableAll))}";
+    };
+    wireServerPackages = (builtins.attrNames (localPackages localModsEnableAll {} {}));
+
+    hoogle = (hPkgs localModsOnlyDocs).hoogleWithPackages (p: builtins.map (e: p.${e}) wireServerPackages);
+
+    # More about dockerTools.streamLayeredImage:
+    # https://nixos.org/manual/nixpkgs/unstable/#ssec-pkgs-dockerTools-streamLayeredImage
+    hoogleImage = pkgs.dockerTools.streamLayeredImage {
+      name = "quay.io/wire/wire-server-hoogle";
+      maxLayers = 50;
+      contents = [
+        hoogle
+        pkgs.coreutils
+        pkgs.bashInteractive
+      ];
+    };
+
+    # Tools common between CI and developers
+    commonTools =  [
+      pkgs.cabal2nix
+      pkgs.gnumake
+      pkgs.gnused
+      pkgs.helm
+      pkgs.helmfile
+      pkgs.hlint
+      pkgs.jq
+      pkgs.kubectl
+      pkgs.ormolu
+      pkgs.shellcheck
+      (hlib.justStaticExecutables pkgs.haskellPackages.cabal-fmt)
+    ] ++ pkgs.lib.optionals pkgs.stdenv.isLinux [
+      pkgs.skopeo
+    ];
+
+    sources = import ./sources.nix;
+    ciImage = import "${sources.nix}/docker.nix" {
+      inherit pkgs;
+      name = "quay.io/wire/wire-server-ci";
+      maxLayers = 2;
+      # We don't need to push the "latest" tag, every step in CI should depend
+      # deterministically on a specific image.
+      tag = null;
+      bundleNixpkgs = false;
+      extraPkgs = commonTools ++ [pkgs.cachix];
+      nixConf = {
+        experimental-features = "nix-command flakes";
```

Comment:
Do we still need this, or is a nix without flakes sufficient?

I probably still can build wire-server without enabling flakes, so i don't se why CI would need to...

Reply:
I'm running Nix 2.3 on my laptop, which predates the introduction of flakes, and Akshay's gone to some effort to ensure that the nix-ified build still works for me. I would think that we don't need the flakes feature flag here, unless there are changes in recent Nix so that features which were previously independent of flakes now depend on them being enabled.